cpp/order/CubaticOrderParameter.cc

```diff
@@ -6,148 +6,175 @@
+    tensor4 c;
+    for (unsigned int i = 0; i < 81; i++)
+    {
+        c.data[i] = data[i] * b;
+    }
+    return c;
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getGlobalTensor()
+void tensor4::copyToManagedArray(util::ManagedArray<float> &ma)
 {
-    memcpy(m_sp_global_tensor.get(), (void*) &m_global_tensor.data, sizeof(float) * 81);
-    return m_sp_global_tensor;
+    memcpy(ma.get(), (void*) &data, sizeof(float) * 81);
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getCubaticTensor()
+//! Complete tensor contraction.
+/*! This function is simply a sum-product over two tensors. For reference, see eq. 4.
+ *
+ *  \param a The first tensor.
+ *  \param a The second tensor.
+ */ 
+float dot(const tensor4& a, const tensor4& b)
 {
-    memcpy(m_sp_cubatic_tensor.get(), (void*) &m_cubatic_tensor.data, sizeof(float) * 81);
-    return m_sp_cubatic_tensor;
+    float c = 0;
+    for (unsigned int i = 0; i < 81; i++)
+    {
+        c += a.data[i] * b.data[i];
+    }
+    return c;
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getGenR4Tensor()
+//! Generate the r4 tensor.
+/*! The r4 tensor is not a word used in the paper, but is a name introduced in
+ *  this code to refer to the second term in eqs. 27 in the paper. It is simply
+ *  a scaled sum of some delta function products. For convenience, its
+ *  calculation is performed in a single function.
+ */ 
+tensor4 genR4Tensor()
 {
-    memcpy(m_sp_gen_r4_tensor.get(), (void*) &m_gen_r4_tensor.data, sizeof(float) * 81);
-    return m_sp_gen_r4_tensor;
-}
+    // Construct the identity matrix to build the delta functions.
+    util::ManagedArray<float> identity({3, 3});
+    identity(0, 0) = 1;
+    identity(1, 1) = 1;
+    identity(2, 2) = 1;
 
-unsigned int CubaticOrderParameter::getNumParticles()
-{
-    return m_n;
+    unsigned int cnt = 0;
+    tensor4 r4 = tensor4();
+    for (unsigned int i = 0; i < 3; ++i)
+        for (unsigned int j = 0; j < 3; ++j)
+            for (unsigned int k = 0; k < 3; ++k)
+                for (unsigned int l = 0; l < 3; ++l)
+                {
+                    // ijkl term
+                    r4[cnt] += identity(i, j)*identity(k, l);
+                    // ikjl term
+                    r4[cnt] += identity(i, k)*identity(j, l);
+                    // iljk term
+                    r4[cnt] += identity(i, l)*identity(j, k);
+                    r4[cnt] *= 2.0/5.0;
+                    ++cnt;
+                }
+    return r4;
 }
 
-float CubaticOrderParameter::getTInitial()
-{
-    return m_t_initial;
-}
 
-float CubaticOrderParameter::getTFinal()
+CubaticOrderParameter::CubaticOrderParameter(float t_initial, float t_final, float scale,
+                                             unsigned int replicates, unsigned int seed)
+    : m_t_initial(t_initial), m_t_final(t_final), m_scale(scale), m_n(0), m_replicates(replicates),
+      m_seed(seed)
 {
-    return m_t_final;
+    // sanity checks, should be caught in python
```

Comment:
Is this still true? If not, remove the comment. I think we should defer our Python checks and just throw C++ exceptions if possible.